### PR TITLE
feat(EC-272) check that operator package name is allowed

### PR DESCRIPTION
**Pull request by ralphbean**
This checks that the FBC pipeline can only publish operator packages which are declared explicitly on the ReleasePlanAdmission.

The purpose of this change is to prevent two different teams from publishing the same operator package name to the same index, stomping on each other. That uniqueness requirement is enforced elsewhere, where we ensure that all RPAs in a managed workspace have no conflicting allowedPackages values.

See also https://gitlab.cee.redhat.com/releng/rhtap-release-data/-/merge_requests/209

### CODE REVIEW

**Comment by ralphbean** [comment]
Here, I still need to figure out how to either get the `opm` binary added to the release-service-utils image or get the task to run using a different image that includes that binary.

File: tasks/check-fbc-packages/check-fbc-packages.yaml
Comment (johnbieren) [inline tasks/check-fbc-packages/check-fbc-packages.yaml]: this should be like 0.1.0 or something

File: tasks/check-fbc-packages/check-fbc-packages.yaml
Comment (johnbieren) [inline tasks/check-fbc-packages/check-fbc-packages.yaml]: maybe this was leftover from the copied task?

File: tasks/check-fbc-packages/check-fbc-packages.yaml
Comment (happybhati) [inline tasks/check-fbc-packages/check-fbc-packages.yaml]:
```suggestion
    app.kubernetes.io/version: "0.1.0"
```

File: tasks/check-fbc-packages/tests/pre-apply-task-hook.sh
Comment (johnbieren) [inline tasks/check-fbc-packages/tests/pre-apply-task-hook.sh]: Don't think this is needed

File: tasks/check-fbc-packages/tests/test-check-fbc-packages-negative.yaml
Comment (johnbieren) [inline tasks/check-fbc-packages/tests/test-check-fbc-packages-negative.yaml]: doesn't match filename

File: tasks/check-fbc-packages/tests/test-check-fbc-packages-positive.yaml
Comment (johnbieren) [inline tasks/check-fbc-packages/tests/test-check-fbc-packages-positive.yaml]: maybe this is still wip

File: tasks/check-fbc-packages/README.md
Comment (johnbieren) [inline tasks/check-fbc-packages/README.md]: this shouldn't be needed

File: pipelines/fbc-release/fbc-release.yaml
Comment (johnbieren) [inline pipelines/fbc-release/fbc-release.yaml]: we replaced these in favor of collect-data task results (this and line 241)

File: pipelines/fbc-release/fbc-release.yaml
Comment (johnbieren) [inline pipelines/fbc-release/fbc-release.yaml:1]: can you bump the y of the version and add a changelog entry?

File: tasks/check-fbc-packages/check-fbc-packages.yaml
Comment (ralphbean) [reply tasks/check-fbc-packages/check-fbc-packages.yaml]: It sure was! Fixing now.

File: tasks/check-fbc-packages/check-fbc-packages.yaml
Comment (happybhati) [reply tasks/check-fbc-packages/check-fbc-packages.yaml]: Resolving, Johhny already mentioned in his review.

File: tasks/check-fbc-packages/tests/test-check-fbc-packages-positive.yaml
Comment (ralphbean) [reply tasks/check-fbc-packages/tests/test-check-fbc-packages-positive.yaml]: No, it's leftover. Just deleting it.

File: tasks/check-fbc-packages/check-fbc-packages.yaml
Comment (ralphbean) [reply tasks/check-fbc-packages/check-fbc-packages.yaml]: Fixed.

File: tasks/check-fbc-packages/tests/test-check-fbc-packages-negative.yaml
Comment (ralphbean) [reply tasks/check-fbc-packages/tests/test-check-fbc-packages-negative.yaml]: Fixed.

File: tasks/check-fbc-packages/README.md
Comment (ralphbean) [reply tasks/check-fbc-packages/README.md]: Fixed.

File: pipelines/fbc-release/fbc-release.yaml
Comment (ralphbean) [reply pipelines/fbc-release/fbc-release.yaml:1]: Yes, done.

File: pipelines/fbc-release/fbc-release.yaml
Comment (ralphbean) [reply pipelines/fbc-release/fbc-release.yaml]: Got it. Fixed!

**Comment by ralphbean** [comment]
> Here, I still need to figure out how to either get the `opm` binary added to the release-service-utils image or get the task to run using a different image that includes that binary.

Oh, it's already there: https://github.com/redhat-appstudio/release-service-utils/blob/main/Dockerfile#L12

**Review by johnbieren — DISMISSED** [review]

**Comment by jinqi7** [comment]
The [RPA](https://github.com/redhat-appstudio/e2e-tests/blob/dc55fcf081b4b8ad859a06c6ec9be598f1458463/tests/release/pipelines/fbc_release.go#L313C1-L332C4) in fbc-release.go of e2e-tests need to change to cooperate with this PR. A paired PR in e2e-tests is needed.

**Comment by johnbieren** [comment]
> The [RPA](https://github.com/redhat-appstudio/e2e-tests/blob/dc55fcf081b4b8ad859a06c6ec9be598f1458463/tests/release/pipelines/fbc_release.go#L313C1-L332C4) in fbc-release.go of e2e-tests need to change to cooperate with this PR. A paired PR in e2e-tests is needed.

Great catch, Jing. Let's get all the fbc RPAs across e2e and users updated before we merge this. It wouldn't break anybody while it is only in development, but I'd like to eliminate the timing variable with promoting it if possible, which is possible here

**Comment by ralphbean** [comment]
Thanks all. Hold on merging this. I'll follow up on the e2e RPA change.

**Comment by ralphbean** [comment]
Posted https://github.com/redhat-appstudio/e2e-tests/pull/1068.

**Comment by johnbieren** [comment]
/test release-pipelines-e2e-suite
The e2e change is merged so CI should pass now. It would still be nice to get the production fbc RPAs for users updated too before we take this in, so that users are not blocked

**Comment by ralphbean** [comment]
Totally. Will definitely merge that before we merge this. Let's wait to see that e2e CI passes first.

File: tasks/check-fbc-packages/check-fbc-packages.yaml
Comment (mmalina) [inline tasks/check-fbc-packages/check-fbc-packages.yaml]: This is a list, so at the very least I would suggest adding `[]` to print each item separately, otherwise the result is a json list:
```
jq -r '.p' <<< '{"p":["a","b"]}'
[
  "a",
  "b"
]
```
```
jq -r '.p[]' <<< '{"p":["a","b"]}'
a
b
```

But also, your check will result in allowing substrings of all allowed packages, e.g. `foo` would be allowed even if only `foo-bar` is in `allowedPackages`.

So I'd suggest doing something like this for the check:
```
for package in ${actual_packages}; do
  if jq -e --arg pkg $package ''.fbc.allowedPackages | index($pkg)' ${DATA_FILE}
  then
    ok
  else
    not ok
```

Comment (mmalina) [reply tasks/check-fbc-packages/check-fbc-packages.yaml]: And now I read the thread about this in Slack and it seems this PR will no longer be needed 🙂

Comment (ralphbean) [reply tasks/check-fbc-packages/check-fbc-packages.yaml]: We reversed course again, and we're going to revive this one!

Comment (ralphbean) [reply tasks/check-fbc-packages/check-fbc-packages.yaml]: Thanks @mmalina. I adopted your suggestions.

**Review by johnbieren — COMMENTED** [review]
lgtm, but still shouldn't merge until production rpas updated

**Comment by ralphbean** [comment]
Ack - production RPAs are updated now, as of: https://gitlab.cee.redhat.com/releng/rhtap-release-data/-/merge_requests/209

**Review by johnbieren — APPROVED** [review]

File: tasks/check-fbc-packages/check-fbc-packages.yaml
Comment (mmalina) [reply tasks/check-fbc-packages/check-fbc-packages.yaml]: Cool, it looks good!

**Review by mmalina — APPROVED** [review]

**Comment by ralphbean** [comment]
@johnbieren I think this is ready to merge whenever you have time.

The dependency https://gitlab.cee.redhat.com/releng/konflux-release-data/-/merge_requests/209 is already merged.

**Review by happybhati — APPROVED** [review]
This branch is out-of-date with the base branch: `development`. @ralphbean, could you please rebase whenever you get a chance?

**Comment by ralphbean** [comment]
Thanks @happybhati. Rebase done.

**Review by happybhati — APPROVED** [review]

**Comment by happybhati** [comment]
I noticed merging is blocked
The base branch requires all commits to be signed. [Learn more about signing commits.](https://docs.github.com/articles/about-gpg/)

(One last thing @ralphbean)  {"-"}

**Comment by ralphbean** [comment]
> I noticed merging is blocked The base branch requires all commits to be signed. [Learn more about signing commits.](https://docs.github.com/articles/about-gpg/)

Apologies. Signed and pushed!

**Comment by johnbieren** [comment]
@jinqi7 can you take at the e2e failure on this pr please?

**Comment by johnbieren** [comment]
CI seems fixed. @ralphbean can you rebase?

**Pull request merged**
